pkg/kubelet/kubelet_node_status.go

```diff
@@ -443,16 +449,16 @@ func (kl *Kubelet) syncNodeStatus() {
 		// This will exit immediately if it doesn't need to do anything.
 		kl.registerWithAPIServer()
 	}
-	if err := kl.updateNodeStatus(); err != nil {
+	if err := kl.updateNodeStatus(initialSource); err != nil {
 		klog.ErrorS(err, "Unable to update node status")
 	}
 }
 
 // updateNodeStatus updates node status to master with retries if there is any
 // change or enough time passed from the last sync.
-func (kl *Kubelet) updateNodeStatus() error {
-	klog.V(5).InfoS("Updating node status")
-	for i := 0; i < nodeStatusUpdateRetry; i++ {
+func (kl *Kubelet) updateNodeStatus(initialSource int) error {
```

Comment:
Do I understand correctly that we call this function every 100ms on each node until the node is ready or nodeReadyGracePeriod (120s) now?

And each call will make up to 5 calls:
* one from lister (so not really a kube-apiserver) call
* one from cache
* one from etcd

If so, this concerns me: if some client has e.g. scale up event and adds 1000 nodes at the same time, does it translate to 10k qps of gets?

This is quite a lot for simply node initialization.

Reply:
Yeah it's "up to" 5 calls, but I expect that most of the updateNodeStatus would end up with only a lister call - then it finds that no update is needed and exits afterwards.

The only case I can think of where it runs over to the apiserver is that after an update following lister get call, the lister doesn't get notified of the update and doesn't refresh its data. 100ms later, this routine reads from lister again, gets stale data, attempts to update again and fails, then runs into the next loop where it reads from apiserver.

I hope that we have a way to manually / explicitly refresh lister data after an update but I didn't find one.

Reply:
Let's try a different approach - https://github.com/kubernetes/kubernetes/pull/113466 so I can call each individual piece of the original syncNodeStatus and manage the lifecycle of the node object on my own. Looks like a better starting point?